ui/src/createAzureClient.ts

```diff
@@ -3,24 +3,73 @@
-    addExtensionUserAgent(client);
-    return client;
+import * as types from '../index';
+import { appendExtensionUserAgent } from "./extensionUserAgent";
+
+export function createAzureClient<T>(
+    clientInfo: { credentials: ServiceClientCredentials; subscriptionId: string; environment: Environment; },
+    clientType: new (credentials: ServiceClientCredentials, subscriptionId: string, options?: types.IMinimumServiceClientOptions) => T): T {
+    return new clientType(clientInfo.credentials, clientInfo.subscriptionId, {
+        acceptLanguage: vscode.env.language,
+        baseUri: clientInfo.environment.resourceManagerEndpointUrl,
+        userAgent: appendExtensionUserAgent
+    });
+}
+
+export function createAzureSubscriptionClient<T>(
+    clientInfo: { credentials: ServiceClientCredentials; environment: Environment; },
+    clientType: new (credentials: ServiceClientCredentials, options?: types.IMinimumServiceClientOptions) => T): T {
+    return new clientType(clientInfo.credentials, {
+        acceptLanguage: vscode.env.language,
+        baseUri: clientInfo.environment.resourceManagerEndpointUrl,
+        userAgent: appendExtensionUserAgent
+    });
+}
+
+export function createGenericClient(clientInfo?: ServiceClientCredentials | { credentials: ServiceClientCredentials; environment: Environment; }): ServiceClient {
+    let credentials: ServiceClientCredentials | undefined;
+    let baseUri: string | undefined;
+    if (clientInfo && 'credentials' in clientInfo) {
+        credentials = clientInfo.credentials;
+        baseUri = clientInfo.environment.resourceManagerEndpointUrl;
+    } else {
+        credentials = clientInfo;
+    }
+
+    return new GenericServiceClient(credentials, <types.IMinimumServiceClientOptions>{
+        baseUri,
+        userAgent: appendExtensionUserAgent
+    });
 }
 
-export function createAzureSubscriptionClient<T extends IAddUserAgent>(
-    clientInfo: { credentials: ServiceClientCredentials; environment: AzureEnvironment; },
-    clientType: new (credentials: ServiceClientCredentials, baseUri?: string, options?: AzureServiceClientOptions) => T): T {
-    const client: T = new clientType(clientInfo.credentials, clientInfo.environment.resourceManagerEndpointUrl, { acceptLanguage: vscode.env.language });
-    addExtensionUserAgent(client);
-    return client;
+class GenericServiceClient extends ServiceClient {
+    constructor(credentials: ServiceClientCredentials | undefined, options: types.IMinimumServiceClientOptions) {
+        super(credentials, options);
+
+        if (options.baseUri) {
+            if (options.baseUri.endsWith('/')) {
+                options.baseUri = options.baseUri.slice(0, -1);
+            }
+            this.baseUri = options.baseUri;
+        }
+    }
+
+    public async sendRequest(options: RequestPrepareOptions | WebResourceLike): Promise<HttpOperationResponse> {
```

Comment:
Is this to replace the requestUtils.sendRequest?  I don't understand why it's in the `createAzureClient` file.

Reply:
It's a part of `GenericServiceClient`, which is returned in `createGenericClient`

Reply:
But anyways - yes this is related to replacing `requestUtils`. I figured `createGenericClient` was similar enough to `createAzureClient` that it could fit here. Only the function `createGenericClient` is actually exported, _not_ `GenericServiceClient`